Basic Multiwallet GUI support

luke-jr opened the pull request:
This adds a combobox to the GUI main window and debug window to select the wallet being viewed/used.

luke-jr committed:
Qt: Load all wallets into WalletModels

luke-jr committed:
Qt: Add a combobox to toolbar to select from multiple wallets

luke-jr committed:
Qt: Add wallet selector to debug console

luke-jr committed:
Qt: Ensure UI updates only come from the currently selected walletView

luke-jr committed:
RPC: Pass wallet through JSONRPCRequest

luke-jr committed:
Qt: QComboBox::setVisible doesn't work in toolbars, so defer adding it at all until needed

luke-jr committed:
Qt: When multiple wallets are used, include in notifications the name of the wallet involved

jonasschnelli commented on src/qt/bitcoin.cpp:
```diff
@@ -484,16 +486,25 @@ void BitcoinApplication::initializeResult(bool success)
-            window->addWallet(BitcoinGUI::DEFAULT_WALLET, walletModel);
-            window->setCurrentWallet(BitcoinGUI::DEFAULT_WALLET);
+            QString WalletName = QString::fromStdString(pwallet->GetName());
+            if (WalletName.endsWith(".dat")) {
```
Hmm.. not sure if we should hardcode `.dat` at this point.

luke-jr replied:
It's strictly for stripping.

ryanofsky replied:
https://github.com/bitcoin/bitcoin/pull/11383#discussion_r140410333

> Hmm.. not sure if we should hardcode .dat at this point.

Note: this is now cleaned up in a later commit ("Get wallet name from WalletModel rather than passing it around")

jonasschnelli commented on src/qt/bitcoingui.cpp:
```diff
@@ -462,6 +462,15 @@ void BitcoinGUI::createToolBars()
         toolbar->addAction(historyAction);
         overviewAction->setChecked(true);
+
+#ifdef ENABLE_WALLET
```
this `#ifdef` could be avoided/removed, right?

luke-jr replied:
Maybe.

jonasschnelli replied:
Has to bee there I guess because of `bool setCurrentWallet(const QString& name);` being only defined if `ENABLE_WALLET`.

jonasschnelli commented:
Nice!
Works pretty good (played around with it a bit).
I think this may be a first step.

Conceptual issues:
* The Combox box without a direct user feedback may lead to selecting the wrong wallet. Android dual/multi-sim works against this by always displaying an icon for the SIM card the action will be executed with. Not 1:1 applicable here... but users should get A) a feedback after switching wallets (are you sure, etc.), B) better visual distinction.
* Having the HD / encryption status in the status-bar seems not to be the right place when running multiwallet. It is basically on the same level then the QComboBox.
* Wallet name tied to the filename may be not the best choice. Would something speak against having a dedicated `walletname` per wallet, .. or would this confuse?

Ideas:
* Ideally we would have a wallet-navigation-drawer at the left side that consume a more prominent space and make it move obvious which wallet is used.
* Critical actions like send-coins should maybe remind the user in the confirmation dialog which wallet it will use.
* Users eventually should choose an (coloured) icon per wallet (give a pre-selection of 12 icons) which then will be always visible on a top window level and or displayed when executing critical actions (showing addresses, sending coins, etc.)

Code:
* I don't see why the RPC auth argument would be necessary at this point (but I respect your comment that we should discuss that in #10615)
* I dislike the `#Ifdef ENABLE_WALLET` clustering. I would reduce it to a minimum and runtime bypass where possible. Especially the #Ifdefs in RPC / http low level code is ugly.

jnewbery commented:
I think we can fairly easily pull out the contentious rpcauth parts of this to make it more palatable for review. I have a branch here: https://github.com/jnewbery/bitcoin/tree/pr11383.1 which removes the first three commits and makes the one change to `rpcconsole.cpp` to make this work without rpcauth. It's also rebased on master since there was a merge conflict.

luke-jr committed:
RPC: Add JSONRPCRequest preparation callbacks

luke-jr commented:
Rebased, eliminated `ENABLE_WALLET` stuff added to non-wallet/GUI code, and removed the dependency on `rpcauth` default wallet stuff.

I did not address @jonasschnelli's GUI comments, because I feel some is best explored separately, as improvements on top of this, while others (confirming wallet changes with a prompt) I think would make the feature annoying to use.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -33,21 +33,34 @@
-static const std::string WALLET_ENDPOINT_BASE = "/wallet/";
-
-CWallet *GetWalletForJSONRPCRequest(const JSONRPCRequest& request)
+void JSONRPCRequestWalletResolver(JSONRPCRequest& jreq, const HTTPRequest& httpreq)
```
In commit "RPC: Pass wallet through JSONRPCRequest"

Multiwallet rpc calls are broken in this commit (and the whole PR) because JSONRPCRequestWalletResolver is not actually registered or called anywhere. Maybe a commit was lost in the rebase? The problem causes multiwallet.py test to fail with "Wallet file not specified"

luke-jr replied:
Fixed

ryanofsky commented on src/qt/bitcoin.cpp:
```diff
@@ -252,7 +252,7 @@ public Q_SLOTS:
 #ifdef ENABLE_WALLET
     PaymentServer* paymentServer;
-    WalletModel *walletModel;
+    std::vector<WalletModel*> walletModels;
```
In commit "Qt: Load all wallets into WalletModels"

Might be simpler to make this a vector of unique_ptr, to avoid need for manual deletions. Also since renaming this member anyway, could follow current naming convenetion (m_wallet_models).

luke-jr replied:
Not sure unique_ptr is the right tool for the job here. Renamed.

ryanofsky commented on src/qt/walletview.cpp:
```diff
@@ -95,13 +95,13 @@ void WalletView::setBitcoinGUI(BitcoinGUI *gui)
 
         // Pass through encryption status changed signals
-        connect(this, SIGNAL(encryptionStatusChanged(int)), gui, SLOT(setEncryptionStatus(int)));
+        connect(this, SIGNAL(encryptionStatusChanged(int)), gui, SLOT(updateWalletStatus()));
```
In commit "Qt: Ensure UI updates only come from the currently selected walletView"

This is silently dropping the `int status` value passed to `encryptionStatusChanged`, and the `int hdEnabled` value passed to `hdEnabledStatusChanged.` Assuming this is intended, you should clean up after this change by deleting the unused parameters from `encryptionStatusChanged` and `hdEnabledStatusChanged` declarations and calls.

ryanofsky commented on src/qt/rpcconsole.cpp:
```diff
@@ -843,8 +865,18 @@ void RPCConsole::on_lineEdit_returnPressed()
+        const int wallet_index = ui->WalletSelector->currentIndex();
+        if (wallet_index > 0) {
+            CWalletRef *ppwallet = (CWalletRef*)ui->WalletSelector->itemData(wallet_index).value<void*>();
+            ppwallet = new CWalletRef(*ppwallet);  // Refcount
```
In commit "Qt: Add wallet selector to debug console"

What does refcount comment mean? Is this a todo suggesting that you want to add refcounting in the future? Should clarify comment or maybe remove it.

luke-jr replied:
Clarified.

ryanofsky commented on src/qt/walletmodel.cpp:
```diff
@@ -65,6 +65,15 @@ WalletModel::~WalletModel()
     unsubscribeFromCoreSignals();
 }
 
+QString WalletModel::getWalletName() const
```
In commit "Qt: When multiple wallets are used, include in notifications the name"

Would be nice to follow this up by removing the `name` parameter passed to `BitcoinGUI::addWallet` to avoid duplicating .dat stripping logic.

luke-jr replied:
Done.

luke-jr committed:
Qt: Get wallet name from WalletModel rather than passing it around

luke-jr commented:
(Note Travis failure is due to the batch request bug fixed in #11277)

jnewbery commented:
#11277 is merged. Can you rebase?

luke-jr commented:
Rebased

jonasschnelli commented on src/httprpc.cpp:
```diff
@@ -18,10 +18,23 @@
+    PrepareJSONRPCRequestCallbacks.connect(preparer);
+}
+
+void UnregisterJSONRPCRequestPreparer(const JSONRPCRequestPreparer& preparer)
```
This gets never called, right?

luke-jr replied:
Right, nor does /wallet/ get unregistered... Fixing

jonasschnelli commented on src/httprpc.cpp:
```diff
@@ -236,6 +251,9 @@ bool StartHTTPRPC()
     // ifdef can be removed once we switch to better endpoint support and API versioning
     RegisterHTTPHandler("/wallet/", false, HTTPReq_JSONRPC);
+
+    void JSONRPCRequestWalletResolver(JSONRPCRequest& jreq, const HTTPRequest& httpreq);
```
use `extern`outside function scope?

luke-jr replied:
Isn't `extern` not needed (nor desirable in our codebase) for function references?

jonasschnelli commented on src/httprpc.cpp:
```diff
@@ -18,10 +18,23 @@
 /** WWW-Authenticate to present with 401 Unauthorized response */
 static const char* WWW_AUTH_HEADER_DATA = "Basic realm=\"jsonrpc\"";
 
+boost::signals2::signal<void (JSONRPCRequest&, const HTTPRequest&)> PrepareJSONRPCRequestCallbacks;
```
Isn't the `HTTPRequest` object unused? Why pack it into the signal, future extensions?

luke-jr replied:
Right

jonasschnelli commented on src/qt/rpcconsole.cpp:
```diff
@@ -304,11 +305,10 @@ bool RPCConsole::RPCParseCommandLine(std::string &strResult, const std::string &
-                                // in Qt, use always the wallet with index 0 when running with multiple wallets
-                                QByteArray encodedName = QUrl::toPercentEncoding(QString::fromStdString(vpwallets[0]->GetName()));
-                                req.URI = "/wallet/"+std::string(encodedName.constData(), encodedName.length());
+                            CWalletRef * const ppwallet = (CWalletRef*)ppwallet_v;
```
looks like we should use a C++11 (shared?) pointer here to avoid the manual memory management.

jonasschnelli reviewed (approved):
Lightly tested ACK: 555eec163b1b62615bf0b49469058eb0be894268

Works as intended.
Visually not very elegant, but an acceptable first step.

promag commented on src/rpc/server.h:
```diff
@@ -45,8 +47,9 @@ class JSONRPCRequest
     bool fHelp;
     std::string URI;
     std::string authUser;
+    CWallet *wallet;
```
Do we want wallet here even an opaque pointer?

luke-jr replied:
? This is where it belongs..

promag reviewed (commented):
Tested, but will look the code. Works as expected.

There are some style issues that could be addressed here or in follow ups:
 - combo and label in the toolbar use different style (font for instance) than the buttons;
 - not sure if the right side is the best place in terms of UX;
 - maybe it should be possible to select wallet from the menu bar (for instance, File -> Wallets -> w1);
 - window titles could have the wallet name too;
 - the combo in the console is a bit lost (this time is not in a toolbar);
 - could show a message in the console when the wallet is changed.

Sjors commented on src/qt/bitcoingui.cpp:
```diff
@@ -555,6 +574,11 @@ void BitcoinGUI::removeAllWallets()
+
+bool BitcoinGUI::multiWallet() const
+{
+    return (WalletSelector->count() > 1);
```
Using a UI element to count the number of wallets doesn't seem ideal.

Sjors replied:
I think a better approach is to give QT access to `std::vector<CWalletRef> vpwallets`, and then when the dropdown box (re-)renders itself, it should query that.

That would also avoid passing walletModel instances around like in `rpcConsole->addWallet(walletModel);`.

jonasschnelli replied:
Agree with @Sjors

promag commented:
@luke-jr friendly ping, rebase needed.

laanwj commented:
As per IRC discussion removing this from high priority for review until review comments get addressed.

luke-jr committed:
Bugfix: RPC: Add missing UnregisterHTTPHandler for /wallet/

luke-jr committed:
GUI: RPCConsole: Log wallet changes

luke-jr commented:
Comments addressed and rebased. Also added a fix to unregister the `"/wallet/"` handler.

Sjors reviewed (commented):
I was able to combine this with #11403 without difficulty, so we can probably safely merge either first.

Sjors commented:
I tried making `m_wallet_models` of `BitcoinApplication` public in hopes of being able to use it from `bitcoingui.cpp`, `bitcoin.cpp` and `rpcconsole.cpp`, but that turned into a mess.

I'm still confused as to what's supposed to be the difference between `qt/bitcoin.cpp` and `qt/bitcoin-gui.cpp`. Why are they both importing `qt/bitcoin-gui.h` and why is there `no qt/bitcoin.h`?

Why is the contents of `vpwallets` copied into `m_wallet_models `?

luke-jr commented:
It's not copied... m_wallet_models has models, not the wallets themselves.

ryanofsky reviewed (commented):
utACK f5aa574c372eba83add1e8d9336621e0e475c793.

This looks great, and is very cleanly implemented and nicely broken up into simple commits. (I think I'm going to look into breaking up my commits this way, just based on how ridiculously easy this was to review.)

I agree with various comments from Jonas, John, and Sjors above and think there's room for a lot of UI improvements in the future, but this PR seems like a nice and simple first start.

> I'm still confused as to what's supposed to be the difference between qt/bitcoin.cpp and qt/bitcoin-gui.cpp. Why are they both importing qt/bitcoin-gui.h and why is there no qt/bitcoin.h?

I replied in IRC, but bitcoin-gui.cpp holds just the BitcoinGUI class, while bitcoin.cpp holds init classes that use BitcoinGUI. The naming and organization could probably be improved, though it seems pretty reasonable. #11625 adds a bitcoin.h file, which isn't necessary right now because bitcoin.cpp contains the main() entry point.

jonasschnelli commented on src/qt/bitcoingui.h:
```diff
@@ -171,6 +175,13 @@ public Q_SLOTS:
+    bool setCurrentWallet(const QString& name);
+
+    /** Set the UI status indicators based on the currently selected wallet.
+    */
```
nit: same line comment?

jonasschnelli commented:
Tested again, works.
I think it's worth to fix...
1. The unused signal (remove it and add it when needed)
2. Fix the `delete ppwallet;` (move to shared pointers or similar?)

luke-jr commented:
> The unused signal (remove it and add it when needed)

What unused signal?

> Fix the delete ppwallet; (move to shared pointers or similar?)

I don't know how to do this with Qt. Note that the original design here had the CWalletRef as a shared pointer type, but that still needed a new/delete for the ref itself.

randolf commented:
utACK.

jonasschnelli commented on src/httprpc.cpp:
```diff
@@ -226,6 +241,8 @@ static bool InitRPCAuthentication()
     return true;
 }
 
+void JSONRPCRequestWalletResolver(JSONRPCRequest&, const HTTPRequest&);
```
Looks after a strange design to forward decelerate this here, which is implemented in `rpcwallet.cpp` with not even a `#ifdef ENABLE_WALLET`

jonasschnelli commented:
I would recommend to remove the signal and the JSONRequest wallet member (the JSON RPC layer should not know what a wallet is) and add the URI parsing logic in `rpcconsole.cpp` (though `WalletModel` and a little refactoring of `GetWalletForJSONRPCRequest`?

luke-jr commented:
That would be ugly. :/

jonasschnelli commented:
> That would be ugly. :/

We should just avoid overlapping layers.
Your current design adds wallet knowhow to `JSONRPCRequest` (`server.h`) and adds a boost signal (which we are trying to get rid of) including a forward declaration in the httpserver of a wallet function.
It looks like you have added this only to allow getting the called wallet from within `rpcconsole.cpp`.

IMO it would be much more straight forward to parse the URI there instead of the current invasive approach.

luke-jr commented:
I assume by parse, you mean fabricate...?

The JSONRPCRequest only holds a pointer. This is a good design, unlike making up and then parsing strings internally.

jonasschnelli commented:
> I assume by parse, you mean fabricate...?

Yes. Since the rpcconsole does already fabricate artificial JSONRPCRequests. There is nothing wrong with that. If you want to avoid the artificial JSONRPCRquest creation, you would need to get rid of it completely.

See my commit that...
* ...remove the unnecessary coupling of server.h and wallet
* ...removes the signal
* ...removes "knowhow" of "JSONRPC" from httpserver
* ...simply continues the JSONRPCRequest fabrication

It also makes this PR more compact (17 additions and 61 deletions.)

https://github.com/jonasschnelli/bitcoin/commit/f8eab48a81653a0488f81ad57116515571b82882

The other thing that needs fixing is the passing around of pure pointers for `CWallet*`.
I think, for now, passing around the wallet name could be sufficient (I don't like that,... but seems better then passing around pure pointers). Wallet names are unique for now.

jonasschnelli commented:
Another tiny conceptual problem is that if one switches the wallet in the main window, it does not automatically switch in the console window (Not sure if it should, but if so, it can be address in a follow up PR).

luke-jr commented:
I fundamentally disagree with that approach. It is a poor design, and assumes the RPC layer and GUI share some common wallet name table. Perhaps they do for now, but it doesn't excuse the bad design.

jonasschnelli commented:
I rewrote this PR to avoid the pointer passing and the signal & wallet-server coupling:
https://github.com/bitcoin/bitcoin/compare/master...jonasschnelli:2018/03/multiwallet?expand=1

If I'm alone with my design concerns, then I'm fine dropping it. :)
Would be good the hear other opinions: @ryanofsky @Sjors @promag

Sjors commented:
@jonasschnelli can you turn that into a PR so it's easier to provide line-by-line feedback? It seems `bool BitcoinGUI::multiWallet()` still counts UI elements (`QComboBox *m_wallet_selector` instances). Can that be avoided?

promag commented:
In the "Sending addresses" and "Receiving addresses" dialogs, it's missing the wallet name. Not sure if the dialog title is the best place.

jonasschnelli commented on src/qt/walletview.cpp:
```diff
@@ -166,7 +169,12 @@ void WalletView::processNewTransaction(const QModelIndex& parent, int start, int
+        wallet = walletModel->getWalletName();
+    }
+
+    Q_EMIT incomingTransaction(date, walletModel->getOptionsModel()->getDisplayUnit(), amount, type, address, label, wallet);
```
Why not remove the `bitcoinGui` member and always send the wallet name?

luke-jr replied:
When not in multiwallet mode, this causes the "Wallet" line to be omitted from the notification popup.

jonasschnelli replied:
Since only `bitcoingui.cpp` listens to the `incomingTransaction` signal, it could be added or not added there. IMO the signal emitter should not decide wether to show the wallet name or not.

jonasschnelli commented:
I really like to merge this,... but I would feel pretty uncomfortable to merge code where we unnecessary glue together the RPC/HTTP layer with the wallet as well as manual pointer memory management... this is something that should be fixed in the initial PR IMO rather the count for follow up PR.

luke-jr commented:
@jonasschnelli This only *reduces/removes* existing glue between the RPC/HTTP and wallet. It doesn't add any.

The manual pointer management is unavoidable with Qt AFAIK, without resorting to worse, ugly hacks like passing it as a string (which will break as soon as we support wallet unloading).

jonasschnelli commented:
> @jonasschnelli This only reduces/removes existing glue between the RPC/HTTP and wallet. It doesn't add any.

Are you referring to this PR? In this PR you are trying to add a `CWallet *wallet;` member to `JSONRPCRequest` which IMO is the wrong direction.

> The manual pointer management is unavoidable with Qt AFAIK, without resorting to worse, ugly hacks like passing it as a string (which will break as soon as we support wallet unloading).

Right now, we do identify wallets in multiwallet by it's filename. IMO identifying wallets by it's name is currently superior to passing around `CWallet` pointer with manual deallocation (`delete`)  (which will break worse in case of dynamic loading/unloading IMO).

jonasschnelli commented:
>> The manual pointer management is unavoidable with Qt AFAIK, without resorting to worse, ugly hacks like passing it as a string (which will break as soon as we support wallet unloading).

> Right now, we do identify wallets in multiwallet by it's filename. IMO identifying wallets by it's name is currently superior to passing around CWallet pointer with manual deallocation (delete) (which will break worse in case of dynamic loading/unloading IMO).

Clarification: I'm only referring to the `RPCConsole` here where we simulate a `JSONRPCRequest` (and therefore need to derive the wallet name anyways to use the the correct URI endpoint).

luke-jr commented:
> Are you referring to this PR? In this PR you are trying to add a CWallet *wallet; member to JSONRPCRequest which IMO is the wrong direction.

Indeed I am. This PR improves the situation by moving the wallet-handling logic out of the RPC/HTTP code, into the wallet-specific module. Having a reference to the wallet for the request, on the request is only reasonable. It is a shortcoming of C++ that it needs to be declared together with the rest of the request class. That could be avoided with a `map<string, void*>` I suppose, but then we need to do manual allocation of any future refcounted pointer - and the only gain in this is extremely minor.

> Right now, we do identify wallets in multiwallet by it's filename.

Only for interfacing to HTTP, an inherently text-based protocol. It makes no sense to do so otherwise, and there are serious shortcomings to this approach. (You could very well end up with a multi-step RPC beginning with one wallet, and finishing in another loaded with the same name after the first was unloaded!)

> IMO identifying wallets by it's name is currently superior to passing around CWallet pointer with manual deallocation (delete) (which will break worse in case of dynamic loading/unloading IMO).

This code doesn't do that. The only manual allocation is of a CWalletRef (that is, it is currently a pointer to a pointer). We could store the CWalletRef directly in place of it, but that would break when CWalletRef becomes a refcounting type. The only way to avoid this would be to find a way to teach QVariant to hold a CWalletRef directly.

jonasschnelli commented:
Closing in favour of now merged #12610
